File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        if (s.Length % 2 == 1)
        {
            return false;
        }

        var stack = new Stack<char>(s.Length);
        for (var i = 0; i < s.Length; i++)
        {
            var current = s[i];
            if (IsOpeningParentheses(current))
            {
                stack.Push(current);
            }
            else
            {
                if (stack.Count == 0)
                {
                    return false;
                }
                var open = stack.Pop();
                if (!Matches(open, current))
                {
                    return false;
                }
            }
        }

        return stack.Count == 0;
    }

    // [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool IsOpeningParentheses(char c)
        => c == '(' || c == '{' || c == '[';

    // [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool Matches(char open, char close)
        => (open == '(' && close == ')') || 
            (open == '[' && close == ']') ||
            (open == '{' && close == '}');
}

File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    _PARENTHESES = {"(" : ")", "{" : "}", "[" : "]"}
    def isValid(self, s: str) -> bool:
        opened_parentheses = []

        for c in s:
            if c in self._PARENTHESES:
                opened_parentheses.append(c)
            elif not opened_parentheses or self._PARENTHESES[opened_parentheses[-1]] != c:
                return False
            else:
                opened_parentheses.pop()
        
        return not opened_parentheses